### PR TITLE
Serialize ft tx

### PR DESCRIPTION
**Pull request by yurikoinaba**
Add support for forward transfers in `createRawTx` and `serializeTx`

### CODE REVIEW

File: lib/transaction.js [hunk 1]
```diff
@@ -383,9 +383,38 @@ function serializeTx(txObj) {
...
+      serializedTx += _buf32.toString('hex');
+      var address = txObj.vft_ccout[i].address;
+
+      if (address.length < 64) {
+        address += "0".repeat(64 - address.length);
+      }
+
+      serializedTx += Buffer.from(address, 'hex').reverse().toString('hex');
+      serializedTx += Buffer.from(txObj.vft_ccout[i].scid, 'hex').reverse().toString('hex');
+      serializedTx += Buffer.from(txObj.vft_ccout[i].mcReturnPubKey, 'hex').toString('hex');
```
Comment (yurikoinaba) [inline lib/transaction.js:410]: It seems that the return address publicKey is expected here, instead of the address

File: lib/transaction.js [hunk 2]
```diff
@@ -341,14 +341,14 @@ function serializeTx(txObj) {
...
-  _buf16.writeUInt16LE(txObj.version, 0);
+  _buf16.writeInt32LE(txObj.version, 0);
 
   serializedTx += _buf16.toString('hex'); // History
 
   serializedTx += zbufferutils.numToVarInt(txObj.ins.length);
   txObj.ins.map(function (i) {
     // Txids and vouts
-    _buf16.writeUInt16LE(i.output.vout, 0);
+    _buf16.writeUInt32LE(i.output.vout, 0);
```
Comment (MarcoOl94) [inline lib/transaction.js:351]: From where did you get this change?

File: lib/transaction.js [hunk 3]
```diff
@@ -383,9 +383,38 @@ function serializeTx(txObj) {
...
+
+    serializedTx += zbufferutils.numToVarInt(0); // Write every vsc
+
+    serializedTx += zbufferutils.numToVarInt(txObj.vft_ccout.length); // Write every vft
+
+    for (var i = 0; i < txObj.vft_ccout.length; i++) {
+      var _buf32 = Buffer.alloc(8); // Satoshis
+
+
+      _buf32.writeUInt32LE(txObj.vft_ccout[i].value & -1, 0);
```
Comment (MarcoOl94) [inline lib/transaction.js:399]: Shouldn't be 64LEBN? https://github.com/HorizenOfficial/bitcore-lib-zen/blob/master/lib/transaction/sidechain.js#L447

File: lib/transaction.js [hunk 4]
```diff
@@ -383,9 +383,38 @@ function serializeTx(txObj) {
...
+
+      serializedTx += _buf32.toString('hex');
+      var address = txObj.vft_ccout[i].address;
+
+      if (address.length < 64) {
```
Comment (MarcoOl94) [inline lib/transaction.js:404]: I think we should just force to have a valid 32 bytes address

File: lib/transaction.js (hunk 1 above)
Comment (MarcoOl94) [reply lib/transaction.js:410]: yes it's the pubKeyHash (20 bytes)

File: src/types.js [hunk 5]
```diff
@@ -31,7 +31,8 @@ export type TXOBJ = {
     value: number,
     address: string,
     n: number,
-    mcReturnAddress: string
+    mcReturnAddress: string,
+    mcReturnPubKey?: string
```
Comment (MarcoOl94) [inline src/types.js:35]: We just have one of this. The data is a mainchain pub key hash but also in the zend this field is called mcReturnAddress

**Review by MarcoOl94 — CHANGES_REQUESTED** [review]

File: src/types.js (hunk 5 above)
Comment (yurikoinaba) [reply src/types.js]: I changed it in the deserialization too

**Review by MarcoOl94 — COMMENTED** [review]
LGTM.
In the zend the params "mcReturnAddress" in the RPC "sc_send" is a mainchain address but in the transaction is serialized its pubkeyhash. If you have the possibility you can think to let the user put the mainchain address when creating the transaction and then convert it into a pubkeyhash before the serialization.

**Review by MarcoOl94 — APPROVED** [review]

**Pull request merged**